server/common/fbs/matrix.py

```diff
@@ -46,74 +50,91 @@ def serialize_matrix(builder, n_rows, n_cols, columns, col_idx):
         arr = np.array(bytearray(as_json, "utf-8"))
+        fields = [arr]
     else:
         if sparse.issparse(arr):
-            arr = arr.toarray()
+            row = arr.row
+            data = arr.data
+            size = np.array([max(arr.shape)], dtype="int")
+            fields = [data, row, size]
         elif isinstance(arr, pd.Series):
-            arr = arr.to_numpy()
-        if arr.dtype != as_type:
-            arr = arr.astype(as_type)
-
-    # serialize the ndarray into a vector
-    if arr.ndim == 2:
-        if arr.shape[0] == 1:
-            arr = arr[0]
-        elif arr.shape[1] == 1:
-            arr = arr.T[0]
-
-    vec = builder.CreateNumpyVector(arr)
-
-    # serialize the typed array table
-    builder.StartObject(1)
-    builder.PrependUOffsetTRelativeSlot(0, vec, 0)
+            if arr.dtype.name == "category":
+                secondary = np.array(bytearray(json.dumps(dict(enumerate(arr.cat.categories))), "utf-8"))
+                arr = arr.cat.codes
+                fields = [arr.to_numpy(), secondary]
+            else:
+                fields = [arr.to_numpy()]
+        else:
+            fields = [arr]
+
+        for i, field_type in zip(range(len(fields)), types):
+            if fields[i].dtype != field_type:
+                fields[i] = fields[i].astype(field_type)
+
+            if fields[i].ndim == 2:
+                if fields[i].shape[0] == 1:
+                    fields[i] = fields[i][0]
+                elif fields[i].shape[1] == 1:
+                    fields[i] = fields[i].T[0]
+
+    vecs = [builder.CreateNumpyVector(field) for field in fields]
+    builder.StartObject(len(vecs))
+    for i, vec in enumerate(vecs):
+        builder.PrependUOffsetTRelativeSlot(i, vec, 0)
     array_value = builder.EndObject()
     return (array_type, array_value)
 
 
 def column_encoding(arr):
     column_encoding_type_map = {
-        # array protocol string:  ( array_type, as_type )
-        np.dtype(np.float64).str: (TypedArray.TypedArray.Float32Array, np.float32),
-        np.dtype(np.float32).str: (TypedArray.TypedArray.Float32Array, np.float32),
-        np.dtype(np.float16).str: (TypedArray.TypedArray.Float32Array, np.float32),
-        np.dtype(np.int8).str: (TypedArray.TypedArray.Int32Array, np.int32),
-        np.dtype(np.int16).str: (TypedArray.TypedArray.Int32Array, np.int32),
-        np.dtype(np.int32).str: (TypedArray.TypedArray.Int32Array, np.int32),
-        np.dtype(np.int64).str: (TypedArray.TypedArray.Int32Array, np.int32),
-        np.dtype(np.uint8).str: (TypedArray.TypedArray.Uint32Array, np.uint32),
-        np.dtype(np.uint16).str: (TypedArray.TypedArray.Uint32Array, np.uint32),
-        np.dtype(np.uint32).str: (TypedArray.TypedArray.Uint32Array, np.uint32),
-        np.dtype(np.uint64).str: (TypedArray.TypedArray.Uint32Array, np.uint32),
+        # array protocol string:  ( array_type, as_type1, as_type2 )
+        # as_type2 is used for flatbuffers with more than one slot
```

Comment:
this comment needs additional clarification.  I also suggest improved naming for the fields (maybe used a NamedTuple to be self-documenting?)